feat: make LiteralPredicate serializable via internal IcebergBaseModel

jaimeferj opened the pull request:
<!--
Thanks for opening a pull request!
-->

<!-- In the case this PR will resolve an issue, please replace ${GITHUB_ISSUE_ID} below with the actual Github issue id. -->
Closes [#2523](https://github.com/apache/iceberg-python/issues/2523)

# Rationale for this change

### Spec alignment
`LiteralPredicate.type` uses the same enum as the REST OpenAPI `LiteralExpression.type`:
`"lt" | "lt-eq" | "gt" | "gt-eq" | "eq" | "not-eq" | "starts-with" | "not-starts-with"`.
Source: OpenAPI spec (LiteralExpression).  

Ref: https://github.com/apache/iceberg/blob/b987e60bbd581d6e9e583107d5a85022261ff0d8/open-api/rest-catalog-open-api.yaml#L2264

## Are these changes tested?
yes

## Are there any user-facing changes?

<!-- In the case of user-facing changes, please add the changelog label. -->

jaimeferj commented:
In the issue https://github.com/apache/iceberg-python/issues/2523 it is said to derive the class from `IcebergBaseModel` which I have not done, but could try on if my solution is not accepted.

Fokko commented on pyiceberg/expressions/__init__.py:
```diff
@@ -732,6 +734,39 @@ def __init__(self, term: Union[str, UnboundTerm[Any]], literal: Union[L, Literal
 
+    # ---- JSON (Pydantic) serialization helpers ----
+
+    class _LiteralPredicateModel(IcebergBaseModel):
```
Thanks for raising this PR @jaimeferj, however, this does not make sense to me. What would be the benefit of another class?

jaimeferj replied:
It was the first approach that I went for to not subclass the generic LiteralPredicate with IcebergBaseModel. I will subclass every specialization of LiteralPredicate from IcebergBaseModel and commit the changes and see if it makes more sense for you.

Fokko replied:
Thanks for your understanding, I think subclassing `IcebergBaseModel` is the right approach :)

Fokko commented on pyiceberg/expressions/__init__.py:
```diff
@@ -725,12 +729,37 @@ def as_bound(self) -> Type[BoundNotIn[L]]:
-class LiteralPredicate(UnboundPredicate[L], ABC):
-    literal: Literal[L]
+class LiteralPredicate(IcebergBaseModel, UnboundPredicate[L], ABC):
+    op: str = Field(
+        default="",
+        alias="type",
+        validation_alias="type",
+        serialization_alias="type",
+        repr=False,
+    )
```
Why not call it `type`:
```suggestion
    type: TypingLiteral["lt-eq", "gt", "gt-eq", "eq", "not-eq", "starts-with", "not-starts-with"] = Field(alias="type")
```

jaimeferj replied:
If you call it directly type, you cannot subclass it directly and use:
- For the father: type: TypingLiteral["lt-eq", "gt", "gt-eq", "eq", "not-eq", "starts-with", "not-starts-with"] = Field(alias="type")
- For the child: type = type: TypingLiteral["eq"] = Field(default="eq")

Because mypy will scream at you saying:

```
1. "type" overrides symbol of same name in class "LiteralPredicate"
     Variable is mutable so its type is invariant
       Override type "Literal['eq']" is not the same as base type "Literal['lt-eq', 'gt', 'gt-eq', 'eq', 'not-eq', 'starts-with', 'not-starts-with']" [reportIncompatibleVariableOverride]
```

jaimeferj commented:
Something fishy that I had to pull in order for tests to pass is putting the attribute `term` as `Term[L]` instead of `UnboundTerm[Any]` as it was in `UnboundPredicate`, father of `LiteralPredicate`. However, that also is triggering mypy since we are changing types from father to child.

The problem is that the earlier implementation was calling `_to_unbound_term` when initializing the instance, however, it does not always return `UnboundTerm` as you can easily see from the implementation:

```Python
def _to_unbound_term(term: Union[str, UnboundTerm[Any]]) -> UnboundTerm[Any]:
    return Reference(term) if isinstance(term, str) else term
```

If term is not `UnboundTerm` nor `str` the output is whatever the input was. For example, as done in test `test_not_equal_to_invert` because it is being initialized with a BoundRefence! If you use in the Pydantic model term as `UnboundTerm[L]`:
```Python
_______________________________________________ test_not_equal_to_invert _______________________________________________

    def test_not_equal_to_invert() -> None:
>       bound = NotEqualTo(
            term=BoundReference(  # type: ignore
                field=NestedField(field_id=1, name="foo", field_type=StringType(), required=False),
                accessor=Accessor(position=0, inner=None),
            ),
            literal="hello",
        )

tests/expressions/test_expressions.py:431: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <[AttributeError("'NotEqualTo' object has no attribute 'term'") raised in repr()] NotEqualTo object at 0x7f8d3405c550>
term = BoundReference(field=NestedField(field_id=1, name='foo', field_type=StringType(), required=False), accessor=Accessor(position=0,inner=None))
literal = 'hello'

    def __init__(self, term: Union[str, UnboundTerm[Any]], literal: Union[L, Literal[L]]):  # pylint: disable=W0621
>       super().__init__(term=_to_unbound_term(term), literal=_to_literal(literal))
E       pydantic_core._pydantic_core.ValidationError: 1 validation error for NotEqualTo
E       term
E         Input should be an instance of UnboundTerm [type=is_instance_of, input_value=BoundReference(field=Nest...(position=0,inner=None)), input_type=BoundReference]
E           For further information visit https://errors.pydantic.dev/2.11/v/is_instance_of

pyiceberg/expressions/__init__.py:753: ValidationError
__________________________________________ test_greater_than_or_equal_invert ___________________________________________

    def test_greater_than_or_equal_invert() -> None:
>       bound = GreaterThanOrEqual(
            term=BoundReference(  # type: ignore
                field=NestedField(field_id=1, name="foo", field_type=StringType(), required=False),
                accessor=Accessor(position=0, inner=None),
            ),
            literal="hello",
        )
```

Should we address this in this PR or delegate it to another issue?

jaimeferj commented on pyiceberg/expressions/__init__.py:
```diff
@@ -743,12 +750,52 @@ def as_bound(self) -> Type[BoundNotIn[L]]:
+    @field_validator("value", mode="before")
+    @classmethod
+    def _convert_value(cls, value: Any) -> Literal[Any]:
+        return _to_literal(value)
```
This error im not sure how to tackle it: 
```
  tests/expressions/test_expressions.py:433: error: Argument "term" to "NotEqualTo" has incompatible type "BoundReference[str]"; expected "str | UnboundTerm[Any]"  [arg-type]
```

The test errors can be easily fixed by:

```suggestion
    value: Literal[L] = Field()
    model_config = ConfigDict(populate_by_name=True, frozen=True, arbitrary_types_allowed=True)

    def __init__(self, term: Union[str, UnboundTerm[Any], BoundReference[Any]], literal: Union[L, Literal[L]]):  # pylint: disable=W0621
        super().__init__(term=_to_unbound_term(term), value=_to_literal(literal))  # pylint: disable=W0621
```

Fokko commented on pyiceberg/expressions/__init__.py:
```diff
@@ -52,8 +52,14 @@
+        return term
+    if isinstance(term, BoundReference):
+        return Reference(term.field.name)
+    raise ValueError(f"Expected UnboundTerm | BoundReference | str, got {type(term).__name__}")
```
There were some faulty tests: https://github.com/apache/iceberg-python/pull/2632/

```suggestion
def _to_unbound_term(term: Union[str, UnboundTerm[Any]]) -> UnboundTerm[Any]:
    return Reference(term) if isinstance(term, str) else term
```

jaimeferj replied:
That PR makes sense, thanks for doing it. Now the LiteralPredicate makes more sense to me.

jaimeferj commented:
Rebased to have the new tests from https://github.com/apache/iceberg-python/pull/2632 and implemented the suggested changes.
All tests passing and linter too

kevinjqliu commented on pyiceberg/transforms.py:
```diff
@@ -120,7 +120,7 @@ def _try_import(module_name: str, extras_name: Optional[str] = None) -> types.Mo
 
 
-def _transform_literal(func: Callable[[L], L], lit: Literal[L]) -> Literal[L]:
+def _transform_literal(func: Callable[[Any], Any], lit: Literal[L]) -> Literal[L]:
```
nit: is this change relevant? i dont see `_transform_literal` used anywhere

jaimeferj replied:
In order to silence this mypy errors:

```
- hook id: mypy
- exit code: 1
  pyiceberg/transforms.py:1049: error: Argument 1 to "_transform_literal" has incompatible type "Callable[[str | None], str | None]"; expected "Callable[[str], str]"  [arg-type]
  pyiceberg/transforms.py:1049: error: Argument 1 to "_transform_literal" has incompatible type "Callable[[bool | None], bool | None]"; expected "Callable[[str], str]"  [arg-type]
  pyiceberg/transforms.py:1049: error: Argument 1 to "_transform_literal" has incompatible type "Callable[[int | None], int | None]"; expected "Callable[[str], str]"  [arg-type]
  pyiceberg/transforms.py:1049: error: Argument 1 to "_transform_literal" has incompatible type "Callable[[float | None], float | None]"; expected "Callable[[str], str]"  [arg-type]
  pyiceberg/transforms.py:1049: error: Argument 1 to "_transform_literal" has incompatible type "Callable[[bytes | None], bytes | None]"; expected "Callable[[str], str]"  [arg-type]
  pyiceberg/transforms.py:1049: error: Argument 1 to "_transform_literal" has incompatible type "Callable[[UUID | None], UUID | None]"; expected "Callable[[str], str]"  [arg-type]
  pyiceberg/transforms.py:1049: error: Argument 1 to "_transform_literal" has incompatible type "Callable[[Decimal | None], Decimal | None]"; expected "Callable[[str], str]"  [arg-type]
  pyiceberg/transforms.py:1049: error: Argument 1 to "_transform_literal" has incompatible type "Callable[[datetime | None], datetime | None]"; expected "Callable[[str], str]"  [arg-type]
  pyiceberg/transforms.py:1049: error: Argument 1 to "_transform_literal" has incompatible type "Callable[[date | None], date | None]"; expected "Callable[[str], str]"  [arg-type]
  pyiceberg/transforms.py:1049: error: Argument 1 to "_transform_literal" has incompatible type "Callable[[time | None], time | None]"; expected "Callable[[str], str]"  [arg-type]
  pyiceberg/transforms.py:1051: error: Argument 1 to "_transform_literal" has incompatible type "Callable[[str | None], str | None]"; expected "Callable[[str], str]"  [arg-type]```

kevinjqliu commented on tests/expressions/test_expressions.py:
```diff
@@ -1199,7 +1218,12 @@ def test_bind_ambiguous_name() -> None:
+_assert_literal_predicate_type(EqualTo("a", "b"))
 assert_type(In("a", ("a", "b", "c")), In[str])
 assert_type(In("a", (1, 2, 3)), In[int])
 assert_type(NotIn("a", ("a", "b", "c")), NotIn[str])
```
nit: should we use `_assert_literal_predicate_type` for these too?

jaimeferj replied:
Sure thing

kevinjqliu commented on pyiceberg/expressions/__init__.py:
```diff
@@ -743,12 +743,18 @@ def as_bound(self) -> Type[BoundNotIn[L]]:
-class LiteralPredicate(UnboundPredicate[L], ABC):
-    literal: Literal[L]
+class LiteralPredicate(IcebergBaseModel, UnboundPredicate[L], ABC):
+    type: TypingLiteral["lt", "lt-eq", "gt", "gt-eq", "eq", "not-eq", "starts-with", "not-starts-with"] = Field(alias="type")
```
nit: could you include this in the PR description so its easily referenced in the future?

jaimeferj replied:
done
